backend/src/main/java/codezap/template/service/facade/TemplateApplicationService.java

```diff
@@ -64,23 +79,34 @@ public FindAllTemplatesResponse findAllBy(
             Long memberId, String keyword, Long categoryId, List<Long> tagIds, Pageable pageable
     ) {
         Page<Template> templates = templateService.findAll(memberId, keyword, categoryId, tagIds, pageable);
-        return makeTemplatesResponse(templates);
+        return makeTemplatesResponse(templates, (template) -> false);
+    }
+
+    public FindAllTemplatesResponse findAllByWithMember(Long memberId, String keyword, Long categoryId,
+            List<Long> tagIds, Pageable pageable, Member loginMember
+    ) {
+        Page<Template> templates = templateService.findAll(memberId, keyword, categoryId, tagIds, pageable);
+        return makeTemplatesResponse(templates, (template -> likesService.isLike(template, loginMember)));
```

Comment:
음 LikePredicate 을 아예 안쓰고 false, likesService.isLike(template, loginMember))만 넘겨줘도 똑같지 않나요...?
짱수가 LikePredicate를 어떻게 사용하게 되었는지를 제가 이해하지 못한 것 같아요ㅠ
설명 좀...

```suggestion
        return makeTemplatesResponse(templates, false);
    }

    public FindAllTemplatesResponse findAllByWithMember(Long memberId, String keyword, Long categoryId,
            List<Long> tagIds, Pageable pageable, Member loginMember
    ) {
        Page<Template> templates = templateService.findAll(memberId, keyword, categoryId, tagIds, pageable);
        return makeTemplatesResponse(templates, likesService.isLike(template, loginMember));
```

Reply:
아래 [코멘트](https://github.com/woowacourse-teams/2024-code-zap/pull/680#issuecomment-2368629133)로 답변 남겨두었어요.

Reply:
짱수~ 코멘트 보고 다시 답변해요

제가 전략패턴을 오해하고 있는 걸 수도 있으니 서로 편하게 이야기 해봅시다.
짱수의 의도는 전략 패턴을 통해 로그인했을 경우, 안했을 경우 좋아요 조회 방식을 적용하고 싶은 것 같아요.
다만 현재 코드상으로 그게 잘 보이지는 않는 것 같아요

제가 생각하기에 어색하다고 느껴지는 이유가 

1. 현재 짱수가 말한 전략 패턴이 사용된 건지 한눈에 파악하기 어려운 것 같아요.
2. 좋아요 전략이 조금만 변경이 되어도  Template 파사드가 변경되어버려요. (전략의 책임이 TemplateApplicationSerivce에 있는 것 같아요)

이전에 이야기가 잠깐 나왔던 것처럼 만약 비회원도 좋아요 기능을 사용하게 된다면 어떨까요?
(비회원의 IP를 사용해 좋아요 기능을 구현한다던지, 브라우저 로컬 스토리지에 좋아요를 저장한다던지 여러 방법이 있겠지만 백엔드가 처리한다고 가정해봅시다)
비회원의 좋아요 전략이 변경되면서 위의 현재 `(template) -> false`가 변경되면서 Template 파사드가 변경되어버려요.

제가 이해한 [전략 패턴](https://refactoring.guru/ko/design-patterns/strategy)은 클라이언트 (여기는 전략을 사용하는 TemplateApplication) 가 내부 로직까지는 몰라도 된다고 생각해요.

따라서 LikePredicate 인터페이스를 작성해주신만큼 각 전략에 대한 구현체를 만들면 어떨까요?

(예시 - 이렇게 하라는 거 아님)
```java
@RequiredArgsConstructor
public class LoginMemberLikeStrategy implements LikeStrategy {

    private final LikesService likesService;

    public boolean isLike(Template template, Member member) {
        return likesService.isLike(template, member);
    }
}

// TemplateApplicationService.java

    public FindAllTemplatesResponse findAllByWithMember(Long memberId, String keyword, Long categoryId,
            List<Long> tagIds, Pageable pageable, Member loginMember
    ) {
        Page<Template> templates = templateService.findAll(memberId, keyword, categoryId, tagIds, pageable);
        return makeTemplatesResponse(templates, loginMember, new LoginMemberLikeStrategy(likesService));
    }

```

Reply:
> 1. 현재 짱수가 말한 전략 패턴이 사용된 건지 한눈에 파악하기 어려운 것 같아요.

이 부분이 문제라면.. 제 구현의 한계라고 생각합니다 🥲 
조금 더 고민을 해 보거나,, 다른 방법을 찾아 볼게요

> 2. 좋아요 전략이 조금만 변경이 되어도 Template 파사드가 변경되어버려요. (전략의 책임이 TemplateApplicationSerivce에 있는 것 같아요)

좋아요 전략 변경에 대해 Service 로직이 변경 되는 것은 문제인가요?
구현체를 생성해서 넣어주는 것도 물론 좋은 방법이라 생각합니다.
다만, 좋아요 전략이 변경되었다는 것은 말 그대로 "사용할 전략" 이 변경 되었다는 것이고, 어떤 전략을 사용할지를 선택하는 `TemplateApplicationSerivce` 의 코드가 변경 되는 것도 문제는 없을 것 같습니다!

다시 한번 말하지만, 몰리가 말한 대로 해당 전략의 구현 클래스를 만드는 것도 충분히 좋은 선택이라 생각해요!
단지 `(template) -> false` 로 표현할 수 있는 전략에 구현체를 만들지 말지는 트레이드 오프의 영역으로 느껴져서 코멘트 남깁니다!

Reply:
> 다만, 좋아요 전략이 변경되었다는 것은 말 그대로 "사용할 전략" 이 변경 되었다는 것이고, 어떤 전략을 사용할지를 선택하는 TemplateApplicationSerivce 의 코드가 변경 되는 것도 문제는 없을 것 같습니다!

저도 같은 의미에서 남긴 코멘트였습니다 ㅎㅎ
다시 한번 말하지만 LoginMemberLikeStrategy 또는 UnLoginMemberLikeStrategy 둘 다 선택은 Service에서 하는 게 맞다고 생각해요. 다만 구현 내용까지 Service가 몰라도 된다는 것이었어요.
짱수는 트레이드 오프로 구현체를 만들지 않는 것을 선택했다면 저도 좋아요~ 현재 크게 복잡한 로직은 아니니까요 😁
다만 변경 가능성을 조금 더 고민해보자는 것이었습니다.

> 이 부분이 문제라면.. 제 구현의 한계라고 생각합니다 🥲

구현은 잘되어 있다고 생각해요. 다만 이름만 Predicate가 아닌, 조금 더 의미를 알 수 있도록 Strategy로만 변경해도 바로 파악할 수 있을 것 같아요 😁

Reply:
LikeChecker 로 변경했서용